cachi2/core/package_managers/yarn/resolver.py

```diff
@@ -0,0 +1,197 @@
+    """Fetch and parse dependency info from the 'yarn info' output.
+
+    This function also performs a validation to ensure that the current
+    yarn project can be processed.
+
+    :raises PackageRejected: if the validation fails.
+    :raises SubprocessCallError: if the 'yarn info' command fails.
+    """
+    cmd = ["info", "-AR", "--json", "--cache"]
+
+    result = run_yarn_cmd(cmd, source_dir)
+
+    packages = [Package.from_info_string(info) for info in result.splitlines()]
+
+    _validate_protocols_in_dependencies(packages)
+    _vet_git_dependencies(packages)
+
+    return [_create_component_from_package(package) for package in packages]
+
+
+def _validate_protocols_in_dependencies(packages: list[Package]) -> None:
+    """Validate that Cachi2 can process all protocols present in the dependencies.
+
+    :raises PackageRejected: if a unprocessable protocol is found.
+    """
+    pass
+
+
+def _vet_git_dependencies(packages: list[Package]) -> None:
+    """Vet a Git dependency if it contains forbidden lifecycle scripts.
+
+    :raises PackageRejected: if a dependency containing a forbidden lifecycle
+    script is found.
+    """
+    pass
+
+
+def _resolve_package_name(package: Package) -> str:
+    """Resolve the package's real name based on its protocol.
+
+    Non-registry deps might have names in their locator string that
+    differ from the one in their package.json file.
+
+    Look at the package.json name for every non-registry dependency.
+    """
+    if package.protocol == "npm":
+        return package.name_in_locator
+
+    if package.protocol in ["file"]:
+        return _resolve_remote_file_dependency_name(package)
+
+    if package.protocol in ["portal", "link", "workspace"]:
+        return _resolve_local_file_dependency_name(package)
+
+    if package.protocol in ["git", "github"]:
+        return _resolve_git_dependency_name(package)
+
+    return "foo"
+    # raise RuntimeError(f"Can't resolve package name for {package.locator}")
+
+
+def _resolve_remote_file_dependency_name(package: Package) -> str:
+    """Download the package and check its name on package.json.
+
+    :raises:
+    """
+    return package.name_in_locator
+
+
+def _resolve_local_file_dependency_name(package: Package) -> str:
+    """Check the name of a local package on its package.json.
+
+    :raises:
+    """
+    return package.name_in_locator
+
+
+def _resolve_git_dependency_name(package: Package) -> str:
+    """Clone the package and check its name on its package.json.
+
+    :raises:
+    """
+    return package.name_in_locator
+
+
+def _create_component_from_package(package: Package) -> Component:
+    """Create a SBOM component from a yarn Package."""
+    name = _resolve_package_name(package)
+
+    return Component(
+        name=name,
+        version=package.version,
+        purl=_generate_purl_for_package(package, name),
+    )
+
+
+def _generate_purl_for_package(package: Package, name: str) -> str:
```

Comment:
purl generation will need to respect YarnRc ([npmRegistryServer](https://yarnpkg.com/configuration/yarnrc#npmRegistryServer) and [npmScopes](https://yarnpkg.com/configuration/yarnrc#npmScopes))

Reply:
I changed the YarnRc class to have this info, but I still need to improve the purl generation part.

Reply:
We can leave that for later, IMO for now the important thing is to give _generate_purl_for_package a yarn_rc param and explain why it needs it